Modules/_uuidmodule.c

```diff
@@ -91,45 +121,1625 @@ py_windows_has_stable_node(void)
+// Bytes 6-7:   time_hi_and_version      (16 bits)
+// Bytes 8-9:   clock_seq_and_variant    (16 bits)
+// Bytes 10-15: node                     (48 bits)
+//
+// Note that the time attributes are only relevant to versions 1, 6 and 7.
+//
+// Version field is located in byte 6; most significant 4 bits:
+//
+// Variant field is located in byte 8; most significant variable bits:
+//   0xxx: Reserved for NCS compatibility
+//   10xx: RFC 4122/9562 (standard)
+//   110x: Reserved for Microsoft compatibility
+//   111x: Reserved for future definition
+
+#define RANDOM_BUF_SIZE 256
+#define MAX_FREE_LIST_SIZE 32
+
+/* State of the _uuid module */
+typedef struct {
+    PyTypeObject *UuidType;
+
+    PyObject *safe_uuid;
+    PyObject *safe_uuid_safe;
+    PyObject *safe_uuid_unsafe;
+    PyObject *safe_uuid_unknown;
+
+    PyObject *uint128_max;
+
+    PyObject *reserved_ncs;
+    PyObject *rfc_4122;
+    PyObject *reserved_microsoft;
+    PyObject *reserved_future;
+
+    // UUID v7 state
+    uint64_t last_timestamp_v7;
+    uint64_t last_counter_v7;
+
+    // We overfetch entropy to speed up successive uuid generations;
+    // this enables 10x peformance boost.
+    uint8_t random_buf[RANDOM_BUF_SIZE];
+    uint64_t random_idx;
+    uint64_t random_last_pid;
+
+    // A freelist for uuid objects -- 15-20% performance boost.
+    uuidobject *freelist;
+    uint64_t freelist_size;
+} uuid_state;
+
+#include "clinic/_uuidmodule.c.h"
+
+/*[clinic input]
+class _uuid.UUID "uuidobject *" "&UuidType"
+[clinic start generated code]*/
+/*[clinic end generated code: output=da39a3ee5e6b4b0d input=84ae6e2089cffd3f]*/
+
+// Forward declarations
+static int from_hex(uuidobject *self, PyObject *hex);
+static int from_bytes_le(uuidobject *self, Py_buffer *bytes_le);
+static int from_int(uuidobject *self, PyObject *int_value, int validate);
+static int from_fields(uuidobject *self, PyObject *fields);
+
+static uint64_t
+uuid_getpid(void) {
+    #if !defined(MS_WINDOWS) || defined(MS_WINDOWS_DESKTOP) || defined(MS_WINDOWS_SYSTEM)
+    return (uint64_t)getpid();
 #else
-    ADD_INT("has_stable_extractable_node", 0);
+    return (uint64_t)GetCurrentProcessId();
 #endif
+}
+
+static inline uuid_state *
+get_uuid_state(PyObject *mod)
+{
+    uuid_state *state = PyModule_GetState(mod);
+    assert(state != NULL);
+    return state;
+}
+
+static inline uuid_state *
+get_uuid_state_by_cls(PyTypeObject *cls)
+{
+    uuid_state *state = (uuid_state *)PyType_GetModuleState(cls);
+    assert(state != NULL);
+    return state;
+}
+
+// Forward declaration
+static PyObject *uuid_from_bytes_array(PyTypeObject *type, uint8_t bytes[16]);
+
+static int
+gen_random(uuid_state *state, uint8_t *bytes, Py_ssize_t size)
+{
+    // Overfetching & caching entropy improves the performance 10x.
+    // There's a precedent with NodeJS doing exact same thing for
+    // improving performance of their UUID implementation.
+
+    // IMPORTANT: callers should have a critical section or a lock
```

Comment:
I would remove the comment and rename the method to `gen_random_lock_held` (or a variation to indicate which lock is held). 

In addition you could add inside the method an assert statement to verify the lock is indeed held. For example for critical sections there is `_Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED`